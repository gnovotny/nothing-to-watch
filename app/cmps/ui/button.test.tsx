import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, expect, it, vi } from 'vitest'
import { Button, buttonVariants } from './button'

describe('Button', () => {
  it('should render with default variant and size', () => {
    render(<Button>Click me</Button>)

    const button = screen.getByRole('button')
    expect(button).toBeInTheDocument()
    expect(button).toHaveTextContent('Click me')
  })

  it('should apply variant classes', () => {
    render(<Button variant='outline'>Outline Button</Button>)

    const button = screen.getByRole('button')
    expect(button).toHaveClass('border')
    expect(button).toHaveClass('border-input')
  })

  it('should apply size classes', () => {
    render(<Button size='sm'>Small Button</Button>)

    const button = screen.getByRole('button')
    expect(button).toHaveClass('rounded-md')
    expect(button).toHaveClass('px-3')
  })

  it('should handle onClick events', async () => {
    const user = userEvent.setup()
    const handleClick = vi.fn()

    render(<Button onClick={handleClick}>Clickable</Button>)

    const button = screen.getByRole('button')
    await user.click(button)

    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('should be disabled when disabled prop is true', () => {
    render(<Button disabled>Disabled Button</Button>)

    const button = screen.getByRole('button')
    expect(button).toBeDisabled()
    expect(button).toHaveClass('disabled:pointer-events-none')
  })

  it('should apply custom className alongside variant classes', () => {
    render(<Button className='custom-class'>Custom</Button>)

    const button = screen.getByRole('button')
    expect(button).toHaveClass('custom-class')
    expect(button).toHaveClass('inline-flex') // Default variant class
  })

  it('should forward ref correctly', () => {
    const ref = vi.fn()

    render(<Button ref={ref}>Button with ref</Button>)

    expect(ref).toHaveBeenCalledWith(expect.any(HTMLButtonElement))
  })

  describe('variants', () => {
    const variants = [
      'default',
      'destructive',
      'outline',
      'secondary',
      'ghost',
      'link',
    ] as const

    variants.forEach((variant) => {
      it(`should match snapshot for ${variant} variant`, () => {
        const { container } = render(
          <Button variant={variant}>{variant} Button</Button>,
        )

        expect(container.firstChild).toMatchSnapshot(
          `button-variant-${variant}`,
        )
      })
    })
  })

  describe('sizes', () => {
    const sizes = ['default', 'sm', 'lg', 'icon'] as const

    sizes.forEach((size) => {
      it(`should match snapshot for ${size} size`, () => {
        const { container } = render(<Button size={size}>{size} Button</Button>)

        expect(container.firstChild).toMatchSnapshot(`button-size-${size}`)
      })
    })
  })

  describe('buttonVariants function', () => {
    it('should return correct classes for different combinations', () => {
      const defaultClasses = buttonVariants()
      expect(typeof defaultClasses).toBe('string')
      expect(defaultClasses).toContain('inline-flex')

      const outlineSmallClasses = buttonVariants({
        variant: 'outline',
        size: 'sm',
      })
      expect(outlineSmallClasses).toContain('border-input')
      expect(outlineSmallClasses).toContain('px-3')
    })
  })
})
