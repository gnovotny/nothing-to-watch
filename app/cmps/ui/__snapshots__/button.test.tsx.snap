// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > sizes > should match snapshot for default size > button-size-default 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
>
  default
   Button
</button>
`;

exports[`Button > sizes > should match snapshot for icon size > button-size-icon 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 w-9"
>
  icon
   Button
</button>
`;

exports[`Button > sizes > should match snapshot for lg size > button-size-lg 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-10 rounded-md px-8"
>
  lg
   Button
</button>
`;

exports[`Button > sizes > should match snapshot for sm size > button-size-sm 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-8 rounded-md px-3 text-xs"
>
  sm
   Button
</button>
`;

exports[`Button > variants > should match snapshot for default variant > button-variant-default 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
>
  default
   Button
</button>
`;

exports[`Button > variants > should match snapshot for destructive variant > button-variant-destructive 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90 h-9 px-4 py-2"
>
  destructive
   Button
</button>
`;

exports[`Button > variants > should match snapshot for ghost variant > button-variant-ghost 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
>
  ghost
   Button
</button>
`;

exports[`Button > variants > should match snapshot for link variant > button-variant-link 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 text-primary underline-offset-4 hover:underline h-9 px-4 py-2"
>
  link
   Button
</button>
`;

exports[`Button > variants > should match snapshot for outline variant > button-variant-outline 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input shadow-sm hover:bg-foreground/10 hover:text-accent-foreground h-9 px-4 py-2"
>
  outline
   Button
</button>
`;

exports[`Button > variants > should match snapshot for secondary variant > button-variant-secondary 1`] = `
<button
  class="pointer-events-auto inline-flex cursor-pointer items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-9 px-4 py-2"
>
  secondary
   Button
</button>
`;
